### ISSUE TITLE
Support external signatures (sigs made on system without TUF installed)

### ISSUE THREAD

**Issue by maltfield** [theupdateframework/tuf#1109]
Please fill in the fields below to submit an issue or feature request.  The
more information that is provided, the better.

**Description of issue or feature request**:

Feature request: please add support for `tuf` to be able to import signatures made on another air-gapped machine that doesn't have `tuf` installed (ie: using `gpg` installed through `apt`)

`tuf` has a bootstrap security problem: it's not possible to securely download and install `tuf` because not all of its dependencies are cryptographically signed. For this reason, it's not safe to install `tuf` on the same machine where my private keys live, so I'd rather preform my signatures on a distinct machine, then copy them to my `tuf` machine.

**Current behavior**:

Currently, all metadata creation & signing takes place on a machine requiring `tuf` to be installed

**Expected behavior**:

Newly supported workflow:

1. TUF produces metadata for new version
1. User copies metdata off `TUF` computer to an air-gapped machine (where the release private keys sit in cold storage)
1. User signs the metadata with their private key on the air-gapped machine
1. User copies the signature from the air-gapped machine to the `TUF` computer

**Comment by lukpueh** [comment]
Thanks for the feature request, @maltfield! Have you checked out the related discussion in https://github.com/theupdateframework/tuf/issues/864? Feel free to comment! :)

**Comment by maltfield** [comment]
I skimmed through that ticket but I was under the impression that it was so that `tuf` could make signatures using external tools. Is that correct?

Whereas I'm asking for the ability to make signatures using external tools (not via `tuf`) and then have `tuf` import those signatures made with external tools (eg `gpg`).

**Comment by lukpueh** [comment]
Yes, that's correct, the issues are not the same, only related, especially in regards to the underlying intention of not storing the private keys on the box that runs TUF.

**Comment by lukpueh** [comment]
For GPG we actually already have all of the tools we need for this [`securesystemslib.gpg.functions.create_signature(content, keyid=None, homedir=None)`](https://github.com/secure-systems-lab/securesystemslib/blob/708e9923a432fdc55391b1c65c3467f7a9385617/securesystemslib/gpg/functions.py#L38) calls the `gpg` command in a subprocess and parses the returned data to generate a TUF-like signature. We might as well parse the signature data from a file that we got from your air-gapped box without first calling gpg.

Also, [this section in the tutorial](https://github.com/theupdateframework/tuf/blob/328e8e758d2cef528cda23f6583ccd6e80b5adac/docs/TUTORIAL.md#excursion-dump-metadata-and-append-signature) shows how to pass TUF metadatda to an external signing tool, and how to add an externally created signature back to TUF metadata.

The main problem why we can't do this right away, is that TUF doesn't support above gpg signatures yet, but that's something we also want to change.

**Comment by lukpueh** [comment]
UPDATE 03/22

In [python-tuf v1.0.0](https://github.com/theupdateframework/python-tuf/blob/v1.0.0/docs/1.0.0-ANNOUNCEMENT.md) signature generation/verification, and public key handling are quite nicely decoupled from general metadata generation and modification via the new [Metadata API](https://theupdateframework.readthedocs.io/en/latest/api/tuf.api.html). Here's what the intended workflow would look like now:  

1. Use [`Metadata.to_bytes()`](https://theupdateframework.readthedocs.io/en/latest/api/tuf.api.metadata.metadata.html#tuf.api.metadata.Metadata.to_bytes) with a [`signer=CanonicalJSONSerializer()`](https://theupdateframework.readthedocs.io/en/latest/api/tuf.api.serialization.html#tuf.api.serialization.json.CanonicalJSONSerializer) argument to get the canonical byte representation of any metadata.
2. Sign bytes with out-of-band mechanism.
3. Parse the signature into a [`Signature`](https://github.com/secure-systems-lab/securesystemslib/blob/692b14dac4d1f6ae31c2655e21d5fecd86bd2fe5/securesystemslib/signer.py#L13) object and add it to the [`Metadata.signatures` dictionary](https://github.com/theupdateframework/python-tuf/blob/develop/tuf/api/metadata.py#L113-L114). (I suggest to take a look at [`SslibSigner`](https://github.com/secure-systems-lab/securesystemslib/blob/692b14dac4d1f6ae31c2655e21d5fecd86bd2fe5/securesystemslib/signer.py#L110) for inspiration, which currently is the standard way of creating signature objects for TUF.)

Also, signing is only part of the story. In order for the signatures to be useful, e.g. for a python-tuf powered client, the corresponding public key needs to be in a TUF metadata compatible format and the signature verification routine needs to be supported. We use [`Key.from_securesystemslib_key`](https://theupdateframework.readthedocs.io/en/latest/api/tuf.api.metadata.supporting.html#tuf.api.metadata.Key.from_securesystemslib_key) and [`Key.verify_signature`](https://theupdateframework.readthedocs.io/en/latest/api/tuf.api.metadata.supporting.html#tuf.api.metadata.Key.verify_signature) for public keys and signatures created with our standard securesystemslib signer.

**Comment by jku** [comment]
I'm going to close this:
* signing on an airgapped system is completely reasonable and I don't think python-tuf prevents doing that in any way
* practically you will need a TUF implementation on this airgapped system (the json canonicalization means that signing TUF metadata is not trivial)
